x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java

```diff
@@ -117,41 +117,122 @@ public void setScorer(Scorable s) throws IOException {
 
     @Override
     public InternalAggregation buildAggregation(long bucket) throws IOException {
-        if (metricValueSource == null) {
-            return buildEmptyAggregation();
-        }
-        List<InternalTopMetrics.TopMetric> topMetrics = sort.getValues(bucket, values.resultBuilder(sort.getFormat()));
+        List<InternalTopMetrics.TopMetric> topMetrics = sort.getValues(bucket, metrics.resultBuilder(sort.getFormat()));
         assert topMetrics.size() <= size;
-        return new InternalTopMetrics(name, sort.getOrder(), metricName, size, topMetrics, pipelineAggregators(), metaData());
+        return new InternalTopMetrics(name, sort.getOrder(), metrics.names, size, topMetrics, pipelineAggregators(), metaData());
     }
 
     @Override
     public InternalTopMetrics buildEmptyAggregation() {
-        // The sort format and sort order aren't used in reduction so we pass the simplest thing.
-        return InternalTopMetrics.buildEmptyAggregation(name, metricName, pipelineAggregators(),
-                metaData());
+        return InternalTopMetrics.buildEmptyAggregation(name, metrics.names, pipelineAggregators(), metaData());
     }
 
     @Override
     public void doClose() {
-        Releasables.close(sort, values);
+        Releasables.close(sort, metrics);
+    }
+
+    private static class Metrics implements BucketedSort.ExtraData, Releasable {
+        private final List<String> names;
+        private final MetricValues[] values;
+
+        Metrics(int size, QueryShardContext ctx, List<MultiValuesSourceFieldConfig> fieldsConfig) {
+            names = fieldsConfig.stream().map(MultiValuesSourceFieldConfig::getFieldName).collect(toList());
+            values = new MetricValues[fieldsConfig.size()];
+            int i = 0;
+            for (MultiValuesSourceFieldConfig config : fieldsConfig) {
+                ValuesSourceConfig<ValuesSource.Numeric> resolved = ValuesSourceConfig.resolve(ctx, ValueType.NUMERIC,
+                        config.getFieldName(), config.getScript(), config.getMissing(), config.getTimeZone(), null);
+                if (resolved == null) {
+                    values[i++] = new MissingMetricValues();
+                    continue;
+                }
+                ValuesSource.Numeric valuesSource = resolved.toValuesSource(ctx);
```

Comment:
Be aware that this can throw on invalid configs (i.e. configs containing neither a script nor a field)